File: Client/game/QuestModel.cpp
/**
 * @file    QuestModel.cpp
 * @brief   QuestModel 实现
 */

#include "game/QuestModel.h"

QuestModel::QuestModel() = default;

void QuestModel::clear()
{
    m_quests.clear();
}

void QuestModel::upsert(const QuestEntry& entry)
{
    for (auto& q : m_quests)
    {
        if (q.questId == entry.questId)
        {
            q = entry;
            return;
        }
    }
    m_quests.push_back(entry);
}

const std::vector<QuestEntry>& QuestModel::quests() const
{
    return m_quests;
}

std::string QuestModel::primarySummary() const
{
    if (m_quests.empty())
    {
        return u8"暂无任务";
    }
    const QuestEntry& q = m_quests.front();
    return q.name + " (" + std::to_string(q.progress) + "/" + std::to_string(q.target) + ")";
}

File: assets/_Project/Scripts/Game/QuestModel.cs
/// <summary>
/// 任务运行时状态。
/// </summary>
using System;
using System.Collections.Generic;

namespace Rpg.Client.Game
{
    public sealed class QuestEntry
    {
        public uint QuestId;
        public string Name = string.Empty;
        public uint Progress;
        public uint Target;
        public bool Done;
    }

    public sealed class QuestModel
    {
        private readonly Dictionary<uint, QuestEntry> _entries = new Dictionary<uint, QuestEntry>();

        public IReadOnlyDictionary<uint, QuestEntry> Entries => _entries;

        /// <summary>数据变更时触发，UI 可订阅以自动刷新。</summary>
        public event Action OnChanged;

        public void Clear()
        {
            _entries.Clear();
            OnChanged?.Invoke();
        }

        public void Upsert(QuestEntry entry)
        {
            if (entry == null)
            {
                return;
            }

            _entries[entry.QuestId] = entry;
            OnChanged?.Invoke();
        }

        public void ReplaceAll(IEnumerable<QuestEntry> entries)
        {
            _entries.Clear();
            if (entries != null)
            {
                foreach (var entry in entries)
                {
                    if (entry == null || entry.QuestId == 0)
                    {
                        continue;
                    }

                    _entries[entry.QuestId] = entry;
                }
            }

            OnChanged?.Invoke();
        }

        /// <summary>移除指定任务（任务完成/放弃后调用）。</summary>
        public bool Remove(uint questId)
        {
            if (_entries.Remove(questId))
            {
                OnChanged?.Invoke();
                return true;
            }

            return false;
        }

        /// <summary>按 QuestId 查找任务实体。</summary>
        public bool TryGetEntry(uint questId, out QuestEntry entry)
        {
            return _entries.TryGetValue(questId, out entry);
        }
    }
}
